src/outline.rs

```diff
@@ -134,6 +134,87 @@ impl<'a> OutlineNodeRef<'a> {
             }
         }
     }
+
+    /// returns the first child node of this node, if any
+    #[must_use]
+    pub fn first_child(self) -> Option<OutlineNodeRef<'a>> {
+        match self {
+            OutlineNodeRef::Chapter { database, node_id } => {
+                let node = &database.outline_result().tree[node_id];
+                if let Some(sref) = OutlineNodeRef::next_assertion(database, node.stmt_address) {
+                    // return statement following this chapter heading, if any
+                    Some(OutlineNodeRef::Statement { database, sref })
+                } else {
+                    // otherwise, return the first child chapter
+                    Some(OutlineNodeRef::Chapter {
+                        database,
+                        node_id: database.outline_result().tree.first_child(node_id)?,
+                    })
+                }
+            }
+            OutlineNodeRef::Statement { .. } => None,
+        }
+    }
+
+    /// Returns the next statement or chapter, in depth-first traversal order
+    #[must_use]
+    pub fn next(self) -> Option<OutlineNodeRef<'a>> {
+        // First attempt to find a child, since we are depth-first
+        self.first_child().or_else(|| self.next_up())
+    }
+
+    /// Returns the next statement or chapter, in depth-first traversal order
+    /// but never diving into this node's children
+    fn next_up(self) -> Option<OutlineNodeRef<'a>> {
+        match self {
+            // In a chapter, the next node is the next chapter
+            OutlineNodeRef::Chapter { database, node_id } => database
+                .outline_result()
+                .tree
+                .next_sibling(node_id)
+                .map_or_else(
+                    || self.parent()?.next_up(),
+                    |node_id| Some(OutlineNodeRef::Chapter { database, node_id }),
+                ),
+            // In a statement,
+            OutlineNodeRef::Statement { database, sref } => {
+                let address = sref.address();
+                OutlineNodeRef::next_assertion(database, address).map_or_else(
```

Comment:
this should just be a match or if let

Reply:
I first made it a `match`, but then clippy asked for an `if let`